Add additional docs on installations and releases. (#86)
sourcegraph/src-cli: README.md

## Installation

> The preferred method of installation is to ask _your_ Sourcegraph instance for the latest compatible version. To do this, replace `https://sourcegraph.com` in the commands below with the address of your instance. If you are running a Sourcegraph version older than 3.12, these URLs will not be available and you must install directly from sourcegraph.com (by running the following commands _exactly_), or from one of the published [releases on GitHub](https://github.com/sourcegraph/src-cli/releases).

Reviewer: While there's a bit more information here, I found the previous wording easier to follow. Specifically I think we've folded together the "preferred" and "exception" cases a little bit too much. I suggest we try separating them explicitly, for example:

~~~suggestion
**For Sourcegraph 3.12 and newer:** The preferred method of installation is to fetch the latest compatible version from _your_ Sourcegraph instance. To do this, replace `https://sourcegraph.com` with the address of _your instance_ in the commands below.

**For Sourcegraph prior to 3.12:** Sourcegraph versions older than 3.12 do not support these installation URLs, and you must install directly from sourcegraph.com by running the commands below _exactly_ as shown.  Alternatively, you may also install one of the published [releases on GitHub](https://github.com/sourcegraph/src-cli/releases).
~~~

Author: Updated.

3.  Travis will automatically perform the release. Once it has finished, **confirm that the curl commands fetch the latest version above**.
4.  Update the `MinimumVersion` constant in the [src-cli package](https://github.com/sourcegraph/sourcegraph/tree/master/internal/src-cli/consts.go).

### Patch releases

> The version recommended by a Sourcegraph instance will be the highest patch version with the same major and minor version as the set minimum. This means that patch versions are reserved solely for non-breaking changes and minor bug fixes. This allows us to dynamically release fixes for older versions of src-cli without having to update the instance.

Reviewer: I think we can simplify this a bit by avoiding passive voice:

```suggestion
A Sourcegraph instance returns the highest patch version with the same major and minor version as the set minimum. Patch versions are reserved solely for non-breaking changes and minor bug fixes. This allows us to dynamically release fixes for older versions of `src-cli` without having to update the instance.
```

I also don't see any definition of "set minimum" anywhere in scope. What I think we are saying is "the same major and minor version as the oldest compatible version" (similar to minimum version selection in Go). Is that correct?

Author: Updated.

> If you are releasing a bug fix or a new feature that is backwards compatible with one of the previous two minor version of Sourcegraph, the changes should be cherry-picked into a patch branch and re-releases with a new patch version. For example, suppose we have the the recommended versions.

Reviewer: As above, and also a typo fix:

```suggestion
To release a bug fix or a new feature that is backwards compatible with one of the previous two minor version of Sourcegraph, cherry-pick the changes into a patch branch and re-release with a new patch version. 

For example, suppose we have the recommended versions:
```

| Sourcegraph version | Recommended src-cli version |
| ------------------- | --------------------------- | 
| `3.x`               | `3.y.z`                     |
| `3.{x-1}`           | `3.{y-1}.w`                 |

> If a new feature is added to a new `3.y.{z+1}` release of src-cli and this change requires only features available in Sourcegraph `3.{x-1}`, then this feature should also be present in a new `3.{y-1}.{w+1}` release of src-cli. Similar logic should follow for Sourcegraph `3.{x-2}` and its recommended version of src-cli.

Reviewer: This might be easier to follow with specific numbers so the reader does not have to evaluate the version arithmetic at read-time.

For example if we use "3.104" and "3.105" for Sourcegraph versions and "3.68.4" and "3.67.18" for recommended versions, this becomes:

> If a new feature is added in the 3.68.5 release of `src-cli` and this change is compatible with Sourcegraph 3.104, then this feature **must** also be present in a new 3.67.19 release of `src-cli`. Similar logic applies for a release of Sourcegraph 3.103 and its recommended version of the `src-cli` tool

…or words to that effect. 

Also: In this suggestion I removed one of the two distinct uses of "feature" (the feature being released vs. features of the existing product) to avert confusion about which feature is the subject. 

I also replaced "should" with "must", since we are making here a rule for ourselves to follow during releases.

> Because a Sourcegraph instance will automatically select the highest patch version, almost all non-breaking changes should increment only the patch version. The process above only necessary if a backwards-compatible change is made _after_ a backwards-incompatible one. If the recommended src-cli version for Sourcegraph `3.{x-1}` was instead `3.y.x` in the example above, there is no additional step required, and the new patch version of src-cli will be available to both Sourcegraph versions.

Reviewer 2: Is the `x` in `3.y.x` intended to match the `x` in `3.x` under the "Sourcegraph version" column?

Author: Updated.

> Because a Sourcegraph instance will automatically select the highest patch version, almost all non-breaking changes should increment only the patch version. The process above only necessary if a backwards-compatible change is made _after_ a backwards-incompatible one. If the recommended src-cli version for Sourcegraph `3.{x-1}` was instead `3.y.v` in the example above, there is no additional step required, and the new patch version of src-cli will be available to both Sourcegraph versions.

Reviewer: Can we remove "almost" here? Are there notable exceptions?

Also: I found it confusing to read all the way through this only to find at the end that we don't have to do this legerdemain except in this condition. I suggest front-loading the condition more clearly, e.g., "If you have made a backward-compatible change _after_ releasing a backward-incompatible change, you must …" or words to that effect.

Author: Front-loaded.

4.  Update the `MinimumVersion` constant in the [src-cli package](https://github.com/sourcegraph/sourcegraph/tree/master/internal/src-cli/consts.go).

### Patch releases

If a backwards-compatible change is made _after_ a backwards-incompatible one, the backwards-compatible one should be re-released to older instances that support it.

A Sourcegraph instance returns the highest patch version with the same major and minor version as `MinimumVersion` as defined in the instance. Patch versions are reserved solely for non-breaking changes and minor bug fixes. This allows us to dynamically release fixes for older versions of `src-cli` without having to update the instance.

To release a bug fix or a new feature that is backwards compatible with one of the previous two minor version of Sourcegraph, cherry-pick the changes into a patch branch and re-releases with a new patch version. 

For example, suppose we have the the recommended versions.

| Sourcegraph version | Recommended src-cli version |
| ------------------- | --------------------------- | 

> | `3.100`             | `3.90.5`                    |

Reviewer: FYI I thought the x, y stuff was fine in the table, I just figured it would be nice for reading the example if it used real numbers. (This is fine too, though)

| `3.99`              | `3.85.7`                    |

> If a new feature is added to a new `3.91.6` release of src-cli and this change requires only features available in Sourcegraph `3.99`, then this feature should also be present in a new `3.85.8` release of src-cli. Similar logic should follow for Sourcegraph `3.98` and its recommended version of src-cli. Because a Sourcegraph instance will automatically select the highest patch version, all non-breaking changes should increment only the patch version. 

Reviewer: I thought the logic was only for one rev +/-, is that not the case?

Author: Oh yes we've updated it. I'll make the change.
